build-tools/xa-prep-tasks/Xamarin.Android.BuildTools.PrepTasks/SystemUnzip.cs

```diff
@@ -0,0 +1,146 @@
+			}
+
+			if (File.Exists (DestinationFolder.ItemSpec)) {
+				Log.LogError ($"{nameof (DestinationFolder)} must be a directory!");
+				return false;
+			}
+
+			SourceEntryGlobParts    = (SourceEntryGlob ?? "*").Split ('/', '\\');
+
+			Directory.CreateDirectory (DestinationFolder.ItemSpec);
+
+			var tempDir = Path.Combine (Path.GetTempPath (), Path.GetRandomFileName ());
+			Directory.CreateDirectory (tempDir);
+			Log.LogMessage (MessageImportance.Low, $"  Extracting into temporary directory: {tempDir}");
+
+			var encoding = string.IsNullOrEmpty (EntryNameEncoding)
+				? null
+				: Encoding.GetEncoding (EntryNameEncoding);
+
+			var tasks = new TTask [SourceFiles.Length];
+			for (int i = 0; i < SourceFiles.Length; ++i) {
+				var td              = tempDir;
+				var sourceFile      = SourceFiles [i].ItemSpec;
+				var relativeDestDir = SourceFiles [i].GetMetadata ("DestDir");
+				var enc             = encoding;
+				var destFolder      = DestinationFolder.ItemSpec;
+				tasks [i] = TTask.Run (() => ExtractFile (td, sourceFile, relativeDestDir, destFolder, enc));
+			}
+
+			TTask.WaitAll (tasks);
+
+			Directory.Delete (tempDir, recursive: true);
+
+			return !Log.HasLoggedErrors;
+		}
+
+		// Ignore CS1998 because there's no async System.IO APIs to use here.
+		// Instead, we're using Tasks so that we can extract multiple files
+		// in parallel via Task.Run() and Task.WaitAll().
+#pragma warning disable 1998
+		async TTask ExtractFile (string tempDir, string sourceFile, string relativeDestDir, string destinationFolder, Encoding encoding)
+		{
+			var tempName = Path.GetRandomFileName ();
+			var nestedTemp = Path.Combine (tempDir, tempName);
+			Directory.CreateDirectory (nestedTemp);
+
+			relativeDestDir = relativeDestDir?.Replace ('\\', Path.DirectorySeparatorChar);
+
+			if (string.Equals (HostOS, "Windows", StringComparison.OrdinalIgnoreCase)) {
+				ZipFile.ExtractToDirectory (sourceFile, nestedTemp, encoding);
+			} else {
+				var start = new ProcessStartInfo ("unzip", $"\"{sourceFile}\" -d \"{nestedTemp}\"") {
+					CreateNoWindow = true,
+					UseShellExecute = false,
+				};
+				Log.LogMessage (MessageImportance.Low, $"unzip \"{sourceFile}\" -d \"{nestedTemp}\"");
+				var p = Process.Start (start);
+				p.WaitForExit ();
+			}
+
+			var entries = GetExtractedSourceEntries (nestedTemp);
+
+			foreach (var entry in entries) {
+				var name    = Path.GetFileName (entry);
+				var destDir = string.IsNullOrEmpty (relativeDestDir)
+					? destinationFolder
+					: Path.Combine (destinationFolder, relativeDestDir);
+				destDir = Path.Combine (destDir, name);
+				foreach (var file in Directory.EnumerateFiles (entry, "*", SearchOption.AllDirectories)) {
+					var relPath = file.Substring (entry.Length + 1);
+					var dest    = Path.Combine (destDir, relPath);
+					Directory.CreateDirectory (Path.GetDirectoryName (dest));
+					Log.LogMessage (MessageImportance.Low, $"mv '{file}' '{dest}'");
+					if (Directory.Exists (entry))
```

Comment:
The case is for pre-existing directories. Assume `sourceFile` contains a `lib` entry, and `{destinationFolder}/lib` already exists. The simple:

    mv $entry $destinationFolder

amounts to

    mv source/lib whatever/lib

Which creates the new directory `whatever/lib/lib`, which is *not* intended. I want to instead "merge" the contents of `source/lib` with `whatever/lib`. This nested `foreach` supports this "merging" behavior.